
```{r}

##Now truly only coral samples, and sample names ~ gulf/reef sites!
setwd("~/holobiont_2020-2021/PCA_19Sep2022")

set.seed(12222)

###load pckgs

library(vctrs)
library(readr)
library(tidyverse)

```

```{r}
##RDA to link upwelling regime's signal local adaptation to physiological outcome in CBASS

library("smartsnp")
library("poppr")    
library("magrittr")
library('vcfR')
library("ggplot2")
library("RColorBrewer")
library("readr")
library("devtools")


vcf_file <- file.path("~/holobiont_2020-2021/PCA_19Sep2022/poppr/coral_onlyFILTER_bcftools_19Sep2022.vcf.gz")

original_vcf <- read.vcfR("~/holobiont_2020-2021/PCA_19Sep2022/poppr/coral_onlyFILTER_bcftools_19Sep2022.vcf.gz")

head(original_vcf)

### with mtORF

pop.data <- read.table("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final_mtorf_FINAL.csv", sep = ",", header = TRUE)

pop.data

```

```{r}

oldwd <- getwd()

setwd("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022")

file.copy(from = vcf_file, to = "./") # Copy the file to the cwd

```

```{r}
system("plink --vcf coral_onlyFILTER_rename_19Sep2022.vcf --double-id --allow-extra-chr --set-missing-var-ids @:# --indep-pairwise 50 10 0.1 --make-bed --out LD_pruned_bed_region")

### 110193 of 116338 variants removed = 6145 var kept
```

```{r}
###New, pruned file

system("plink --vcf coral_onlyFILTER_rename_19Sep2022.vcf --double-id --allow-extra-chr --set-missing-var-ids @:# --extract LD_pruned_bed_region.prune.in --make-bed --out coral_pruneddata") 

```

```{r}

###Pruned output to vcf


system("plink --bfile coral_pruneddata --double-id --allow-extra-chr --set-missing-var-ids @:# --recode vcf --out vcf_pruned_final_13Feb2024")

```

```{r}

### Pruned to traw

### --recode A-transpose --double-id --allow-extra-chr --out region_genotypeMatrix 

system("plink --bfile coral_pruneddata --recode A-transpose --double-id --allow-extra-chr --set-missing-var-ids @:# --out matrix_vcf_pruned_final_13Feb2024")

```

```{r}
# Load the PLINK (.fam) file to get the number of samples
numSamples = nrow(read.table("coral_pruneddata.fam"))

numSamples
```
```{r}
##Set group variables

pop.data

```

```{r}

numSamples

my_groups <- c(data.table::fread("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final_mtorf_FINAL.csv", header = TRUE))[[4]]

my_groups


my_groups_MLG <- c(data.table::fread("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final_mtorf_FINAL.csv", header = TRUE))[[5]]

my_groups_MLG


my_groups_orf <- c(data.table::fread("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final_mtorf_FINAL.csv", header = TRUE))[[6]]

my_groups_orf
```

```{r}
length(table(my_groups)) #number of reefs

table(my_groups)

length(table(my_groups_MLG)) #number of MLGs

table(my_groups_MLG)

length(table(my_groups_orf)) #number of mtORFs

table(my_groups_orf)
```

```{r}

# Running smart_pca
reef.sm.pca <- smart_pca(snp_data = "matrix_vcf_pruned_final_13Feb2024.traw", 
                    sample_group = my_groups,
                    missing_value = NA, 
                    scaling = "drift",
                    pc_axes = 6)
```
```{r}
reef.sm.pca
```

```{r}
# Here is a plot of the first two components:
plot(reef.sm.pca$pca.sample_coordinates[,c("PC1","PC2")])
```


```{r}
##inspect sm.pca object

class(reef.sm.pca)

names(reef.sm.pca)

str(reef.sm.pca)

```

```{r}
pcaR_eigen <- reef.sm.pca$pca.eigenvalues # extract eigenvalues (PCA1 and PC2 axes explain 3.5% variation in SNP variation across individuals)

pcaR_eigen

pcaR_load <- reef.sm.pca$pca.snp_loadings # extract principal coefficients (high SNP loadings indicate loci with stronger variation across individuals)

pcaR_load 

pcaR_coord <- reef.sm.pca$pca.sample_coordinates # extract principal components (position of individuals in PCA space used to generate the ordination)

pcaR_coord
```

```{r}
cols <- c("#DE1BEB", "#EB1BA3", "#1ED2EB", "#B9F4FF", "#20A4E6", "#EF94EC")
                
```

```{r}
plot(reef.sm.pca$pca.sample_coordinates[,c("PC1","PC2")], cex = 1.5, 
     bg = cols[as.factor(my_groups)], pch = 21, col = "black", main = "Panama TEP WGBS smartpca, by reef")
legend("topright", legend = levels(as.factor(my_groups)), cex = 1, pch = 21, 
       pt.cex = 1.0, col = "black", pt.bg = cols, text.col = cols)

plot(reef.sm.pca$pca.sample_coordinates[,c("PC1","PC3")], cex = 1.5, 
     bg = cols[as.factor(my_groups)], pch = 21, col = "black", main = "Panama TEP WGBS smartpca, by reef")
legend("topright", legend = levels(as.factor(my_groups)), cex = 1, pch = 21, 
       pt.cex = 1.0, col = "black", pt.bg = cols, text.col = cols)

plot(reef.sm.pca$pca.sample_coordinates[,c("PC1","PC4")], cex = 1.5, 
     bg = cols[as.factor(my_groups)], pch = 21, col = "black", main = "Panama TEP WGBS smartpca, by reef")
legend("topright", legend = levels(as.factor(my_groups)), cex = 1, pch = 21, 
       pt.cex = 1.0, col = "black", pt.bg = cols, text.col = cols)
```

```{r}
gulf <- c(data.table::fread("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final.csv", header = TRUE))[[2]]

gulf
```
```{r}

length(table(gulf)) #number of gulfs

table(gulf)
```

```{r}

# Running smart_pca
gulf.sm.pca <- smart_pca(snp_data = "region_genotypeMatrix.traw", 
                    sample_group = gulf,
                    missing_value = NA,
                    scaling = "drift")

# Here is a plot of the first two components:
plot(gulf.sm.pca$pca.sample_coordinates[, c(3,4)])
```

```{r}

##inspect gulf.sm.pca object

class(gulf.sm.pca)

names(gulf.sm.pca)

str(gulf.sm.pca)

```

```{r}
pcaG_eigen <- gulf.sm.pca$pca.eigenvalues 

pcaG_eigen

# extract eigenvalues (PCA1 and PC2 axes explain 3.5% variation in SNP variation across individuals)


pcaG_load <- gulf.sm.pca$pca.snp_loadings 

pcaG_load 

# extract principal coefficients (high SNP loadings indicate loci with stronger variation across individuals)


pcaG_coord <- gulf.sm.pca$pca.sample_coordinates 

pcaG_coord 

# extract principal components (position of individuals in PCA space used to generate the ordination)
```

```{r}
##cols <- rainbow(length(table(gulf)))


cols <- c("#FF33FF", 
                "#56B4E9")
                
plot(gulf.sm.pca$pca.sample_coordinates[,c("PC1","PC2")], cex = 1.5, 
     bg = cols[as.factor(gulf)], pch = 21, col = "black", main = "Panama TEP WGBS smartpca, by gulf")
legend("topright", legend = levels(as.factor(gulf)), cex = 1, pch = 21, 
       pt.cex = 1.0, col = "black", pt.bg = cols, text.col = cols)
```

```{r}
### Does MLG and/or mtORF explain some of these trends above, particularly the seemingly fixed value of PC2?

# Running smart_pca
MLG.sm.pca <- smart_pca(snp_data = "region_genotypeMatrix.traw", 
                    sample_group = my_groups_MLG,
                    missing_value = NA,
                    scaling = "drift")

# Here is a plot of the first two components:
plot(MLG.sm.pca$pca.sample_coordinates[, c(3,4)])

##inspect gulf.sm.pca object

class(MLG.sm.pca)

names(MLG.sm.pca)

str(MLG.sm.pca)
```

```{r}
pcaM_eigen <- MLG.sm.pca$pca.eigenvalues 

pcaM_eigen

# extract eigenvalues (PCA1 and PC2 axes explain 3.5% variation in SNP variation across individuals)

pcaM_load <- MLG.sm.pca$pca.snp_loadings 

pcaM_load 

# extract principal coefficients (high SNP loadings indicate loci with stronger variation across individuals)

```

```{r}
pcaM_coord <- MLG.sm.pca$pca.sample_coordinates 

pcaM_coord 

# extract principal components (position of individuals in PCA space used to generate the ordination)

##cols <- rainbow(length(table(gulf)))

```

```{r}

###Plot smartPCA per MLG

cols <- c("#7FFFD4", "#AAFF00", "#097969", "#00FFFF", "#228B22", "#7CFC00", "#808000", "#0FFF50", "#008080", "#C4B454")
                
plot(MLG.sm.pca$pca.sample_coordinates[,c("PC1","PC2")], cex = 1.5, 
     bg = cols[as.factor(my_groups_MLG)], pch = 21, col = "black", main = "Panama TEP WGBS smartpca, by MLG")
legend("topright", legend = levels(as.factor(my_groups_MLG)), cex = 1, pch = 21, 
       pt.cex = 1.0, col = "black", pt.bg = cols, text.col = cols)
```



```{r}
### Now ... does mtORF explain some of these trends above, particularly the seemingly fixed value of PC2?

# Running smart_pca
orf.sm.pca <- smart_pca(snp_data = "region_genotypeMatrix.traw", 
                    sample_group = my_groups_orf,
                    missing_value = NA,
                    scaling = "drift")

# Here is a plot of the first two components:
plot(orf.sm.pca$pca.sample_coordinates[, c(3,4)])

##inspect gulf.sm.pca object

class(orf.sm.pca)

names(orf.sm.pca)

str(orf.sm.pca)
```

```{r}
pcaO_eigen <- orf.sm.pca$pca.eigenvalues 

pcaO_eigen

# extract eigenvalues (PCA1 and PC2 axes explain 3.5% variation in SNP variation across individuals)

pcaO_load <- orf.sm.pca$pca.snp_loadings 

pcaO_load 

# extract principal coefficients (high SNP loadings indicate loci with stronger variation across individuals)

```

```{r}
pcaO_coord <-orf.sm.pca$pca.sample_coordinates 

pcaO_coord 

# extract principal components (position of individuals in PCA space used to generate the ordination)

##cols <- rainbow(length(table(gulf)))

```

```{r}

###Plot smartPCA per MLG

cols <- c("#FFC300", "#581845")
                
plot(MLG.sm.pca$pca.sample_coordinates[,c("PC1","PC2")], cex = 1.5, 
     bg = cols[as.factor(my_groups_orf)], pch = 21, col = "black", main = "Panama TEP WGBS smartpca, by MLG")
legend("topright", legend = levels(as.factor(my_groups_orf)), cex = 1, pch = 21, 
       pt.cex = 1.0, col = "black", pt.bg = cols, text.col = cols)
```
```{r}
###Let's see who or what is driving differences across the two PCs via histograms

##PC2, reefsite

###First, extract only sample coord into own obj

PCA_reef <- reef.sm.pca$pca.sample_coordinates

PCA_reef

```

```{r}

p1_reef <- ggplot(PCA_reef, aes(x=PC1, color = Group, fill = Group)) + geom_histogram()
p1_reef +  geom_rug()

p2_reef <- ggplot(PCA_reef, aes(x=PC2, color = Group, fill = Group)) + geom_histogram()
p2_reef +  geom_rug()
```

```{r}
## Repeat with gulf

PCA_gulf <- gulf.sm.pca$pca.sample_coordinates

PCA_gulf
```

```{r}

p1_gulf <- ggplot(PCA_gulf, aes(x=PC1, color = Group, fill = Group)) + geom_histogram()
p1_gulf +  geom_rug()

p2_gulf <- ggplot(PCA_gulf, aes(x=PC2, color = Group, fill = Group)) + geom_histogram()
p2_gulf +  geom_rug()
```
```{r}
PCA_orf <- orf.sm.pca$pca.sample_coordinates

PCA_orf

PCA_orf$Group <- as.factor(PCA_orf$Group)

PCA_orf
```

```{r}

p1_orf <- ggplot(PCA_orf, aes(x=PC1, color = Group, fill = Group)) + geom_histogram()
p1_orf +  geom_rug()

p2_orf <- ggplot(PCA_orf, aes(x=PC2, color = Group, fill = Group)) + geom_histogram()
p2_orf +  geom_rug()
```
```{r}
PCA_MLG <- MLG.sm.pca$pca.sample_coordinates

PCA_MLG

PCA_MLG$Group <- as.factor(PCA_MLG$Group)

PCA_MLG
```
```{r}

p1_MLG <- ggplot(PCA_MLG, aes(x=PC1, color = Group, fill = Group)) + geom_histogram()
p1_MLG +  geom_rug()

p2_MLG <- ggplot(PCA_MLG, aes(x=PC2, color = Group, fill = Group)) + geom_histogram()
p2_MLG +  geom_rug()
```

```{r}
gulf <- c(data.table::fread("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final.csv", header = TRUE))[[2]]

permanova_gulf <- smart_permanova(snp_data = "region_genotypeMatrix.traw", sample_group = gulf, 
                              target_space = "pca",  missing_value = NA, scaling = "drift", pairwise = "TRUE", permutation_seed = 1234)

perm_gulf <- permanova_gulf$permanova.samples

permanova_gulf$permanova.global_test

head(permanova_gulf$permanova.pairwise_test)
```

```{r}

my_groups <- c(data.table::fread("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final.csv", header = TRUE))[[4]]

my_groups

permanova_reef <- smart_permanova(snp_data = "region_genotypeMatrix.traw", sample_group = my_groups, missing_value = NA, scaling = "drift", target_space = "pca", pairwise = "TRUE", permutation_seed = 1234)
```

```{r}
perm_reef <- permanova_reef$permanova.samples

permanova_reef$permanova.global_test

head(permanova_reef$permanova.pairwise_test)
```
```{r}

my_groups_orf <- c(data.table::fread("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final_mtorf_FINAL.csv", header = TRUE))[[6]]

my_groups_orf
```

```{r}
my_groups_orf <- na.omit(my_groups_orf)

my_groups_orf

```

```{r}
### For me, from looking at the PC, given the complex genetics and reproduction with TEP corals, outlier detection may best present shared ancestry and bette rinsights into what is driving differences within our highly admixed, low gen div region

###pcadapt may be preferred to outFLANK and BayeScan, among other trad approaches, as does not assume hierarchical pop structure nor require us to define pop, as at least for our system this is arbitrary perhaps

###pcadapt assumes that markers, here SNPs, which are strongly correlated with pop structure, here based on PCs, = candidates for local adaptation

###Additionally, powerful stat for genome scans, the Mahalanobis distance: vector of K z-scores to measure the degree of relatedness to PCs. The Mahalanobis distance is then computed for each SNP to detect outliers for which 78 the vector of z-scores do not follow the distribution of the main bulk of points -- mean/cov of matrix z not sensitive to outliers
```

```{r}
###Running pcadapt

library(pcadapt)
library(qvalue)


###Let's use bed made before with LD prunning, "region"
```

```{r}

vcf.path = "C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\vcf_pruned_final_13Feb2024.vcf"
##bed.path = "C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\region.bed"
meta.path = "C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final_mtorf_FINAL.csv"

genos <- read.pcadapt(vcf.path,type=c("vcf"))

head(genos)
```

```{r}
meta <- read.csv(meta.path)
head(meta)

meta$MLG <- as.factor(meta$MLG)

meta$mtorf <- as.factor(meta$mtorf)

meta
```

```{r}
x <- pcadapt(input=genos,K=6)
plot(x,option="screeplot")

##Chose top 2 PCs as encompass most variance
```

```{r}

library(ggplot2)

##PC per reef
PCA_reef <- plot(x, option="scores",pop=meta$reef)

PCA_reef


cols <- c("#FF33FF", 
                "#56B4E9")

##PC per gulf
PCA_gulf <- plot(x,option="scores",pop=meta$loc, col = cols)


PCA_gulf+theme_bw()

summary(x)

###prop var explained

sing.value <- x$singular.values

sing.value

var_exp <- sing.value^2

var_exp
```

```{r}
PC1 <- x$scores[,1]

PC1
```

```{r}
##PC per mtORF
plot(x,option="scores",pop=meta$mtorf)
```
```{r}
##PC per gulf & mtORF

cols_orf <- c("orange", 
                "purple")
PCA_orf <- plot(x,option="scores",pop=meta$mtorf, col = cols_orf)

PCA_orf + facet_wrap(~meta$loc) + theme_bw()
```

```{r}
##PC1 vs 3 per reef
plot(x, i = 1, j = 3, option="scores",pop=meta$reef)

##PC1 vs 3 per gulf
plot(x, i = 1, j = 3, option="scores",pop=meta$loc)

##PC1 vs 3 per mtORF
plot(x, i = 1, j = 3, option="scores",pop=meta$mtorf)
```

```{r}

##PC1 vs 4 per reef
plot(x, i = 1, j = 4, option="scores",pop=meta$reef)

##PC1 vs 4 per gulf
plot(x, i = 1, j = 4, option="scores",pop=meta$loc)

##PC1 vs 4 per mtORF
plot(x, i = 1, j = 4, option="scores",pop=meta$mtorf)
```

```{r}
## So while PC1 vs PC3 to some degree recaps PC1 vs PC3, at PC1 vs PC4 patterns of BAD in own cluster dissapear
### What does this mean? PC1 ~ 60% var, while PC2 ~ 10%. PC1 in this case comes to represent perhaps a more normal signal, admixture/neutral/standing genetic variation, while PC>1 show exceptionally different pattern up to PC3.

### Loadings of ~0.6 are significant, here PC1

#### Yet in some cases, e.g. UMAP, higher PCs are more likely to represent subtle, but biologically relevant, sources of heterogeneity â€“ so including them may improve downstream analysis

#### For us, this comes to represent how BAD being different from the other sites is biologically relevant if we see this trend continue to emerge even as increase PC1 > ; likewise PC2 ~ PC3 ~ PC4 so the trend seen not an artifact 

### PC1 = largest eigenvalue which captures the maximum variance data points ; these eigenvalues calculated on the covariance matrix

###  To put it in simpler terms, each PC is a direction in your feature space that's highly relevant to determine what the output is. Each PC is less important than the last, as each eigenvector represents less variance.


```


```{r}
plot(x,option="manhattan")
```

```{r}
qval_full <- qvalue(x$pvalues)$qvalues
outliers_full <- which(qval_full<0.1)
length(outliers_full)

qval_full_0.05 <- qvalue(x$pvalues)$qvalues
outliers_full_0.05 <- which(qval_full_0.05 <0.05)
length(outliers_full_0.05)
```

```{r}
hist(x$pvalues,xlab="p-values", main=NULL, breaks=50)

plot(x, option="stat.distribution")
```



```{r}
##Let's re-run with only 2 PC as encompass most variation

x_PC2 <- pcadapt(input=genos,K=2)

summary(x_PC2)
```

```{r}

##Vis outliers

plot(x_PC2 , option = "manhattan")

```
```{r}
##qqplot

plot(x_PC2, option = "qqplot")

```

```{r}
hist(x_PC2$pvalues,xlab="p-values", main=NULL, breaks=50)

plot(x_PC2, option="stat.distribution")
```

```{r}

###Outliers

qval_full <- qvalue(x_PC2$pvalues)$qvalues
outliers_full <- which(qval_full<0.1)
length(outliers_full)

qval_full_0.05 <- qvalue(x_PC2$pvalues)$qvalues
outliers_full_0.05 <- which(qval_full_0.05 <0.05)
length(outliers_full_0.05)
```

```{r}
vcfR_pruned<-read.vcfR("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\vcf_pruned_final_13Feb2024.vcf")

vcfR_pruned
```

```{r}

vcf_subset_FDR0.05 <- vcfR_pruned[outliers_full_0.05, ]

head(vcf_subset_FDR0.05)
```

```{r}

write.vcf(vcf_subset_FDR0.05 , "q0.05pcadapt_outliersVCF_TEP_corals_13Feb2024.vcf")
```

```{r}
head(vcf_subset_FDR0.05)
```

```{r}
vcf_subset_FDR0.05 <- read.vcfR("q0.05pcadapt_outliersVCF_TEP_corals_13Feb2024.vcf")
```

```{r}
set.seed(1555)
###Let's re-do the pruned loci but with unique sample names

system("plink --vcf coral_onlyFILTER_bcftools_19Sep2022.vcf --double-id --allow-extra-chr --set-missing-var-ids @:# --indep-pairwise 50 10 0.1 --make-bed --out uniquenames_LD_pruned_bed_region")

### 110193 of 116338 variants removed = 6145 var kept
```

```{r}
###New, pruned file

system("plink --vcf coral_onlyFILTER_bcftools_19Sep2022.vcf --double-id --allow-extra-chr --set-missing-var-ids @:# --extract uniquenames_LD_pruned_bed_region.prune.in --make-bed --out uniquenames_coral_pruneddata") 

```

```{r}

###Pruned output to vcf


system("plink --bfile uniquenames_coral_pruneddata --double-id --allow-extra-chr --set-missing-var-ids @:# --recode vcf --out vcf_uniquenames_coral_pruneddata_final_13Feb2024")

```


```{r}
##Final VCF, without rename to have unique ids; LD pruned 

vcf_prunedLD_uniquenames <- read.vcfR("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\vcf_uniquenames_coral_pruneddata_final_13Feb2024.vcf")

head(vcf_prunedLD_uniquenames)

pop.data <- read.table("C:\\Users\\vmgly\\Documents\\holobiont_2020-2021\\PCA_19Sep2022\\population.gbs_final_GenAl_3Oct2022.csv", sep = ",", header = TRUE)

head(pop.data)
```

```{r}

genind <- vcfR2genind(vcf_prunedLD_uniquenames) ##keep this one without pop defined

genind_gulf <- vcfR2genind(vcf_prunedLD_uniquenames)
```

```{r}
genind_gulf

genind_gulf@pop
```

```{r}
###Add gulf information 

genind_gulf@pop <- as.factor(pop.data$loc)

head(genind_gulf)
```

```{r}
##see pop

genind_gulf@pop
```

```{r}
library(hierfstat)

basic.stats(genind_gulf, diploid = TRUE)

basic.stats
```

```{r}
wc(genind_gulf) # Weir and Cockerham's estimate
```

```{r}

####Per gulf pairwise Fst
##coral_fst = genet.dist(genind, method = "WC84") %>% round(digits = 3)
##coral_fst
```

```{r}
### Okay re-read vcf to do pairwise gulf comparison

genind_reef <- vcfR2genind(vcf_prunedLD_uniquenames)

genind_reef@pop
```

```{r}
genind

head(vcf_prunedLD_uniquenames)
```

```{r}

###make sure sample IDs match vcf and pop.data

pop.data

pop.data2 <- pop.data[order(pop.data$DNA_ID), ]

pop.data2
```

```{r}
reef_strata <- pop.data2$reef

reef_strata
```


```{r}
###Add gulf information 

genind_reef@pop <- as.factor(pop.data2$reef)

genind_reef@pop

head(genind_reef)
```

```{r}

head(genind)
```

```{r}

coral_reef_fst = genet.dist(genind_reef, method = "WC84") %>% round(digits = 3)
coral_reef_fst
```

```{r}
###Let's do this from first principle the outlier detection:


library(hierfstat)
library(MASS)


##Subset all pairs of pop

# Choose the populations you want to keep (replace "pop1" and "pop2" with your desired population names)

##Gulf of Panama

pop_S_M <- c("SAB_A", "MOG_A")

pop_S_C <- c("SAB_A", "CON_A")

pop_M_C <- c("MOG_A", "CON_A")

##Gulf of Chiriqui 

pop_B_A <- c("BAD_A", "AFU_A")

pop_B_U <- c("BAD_A", "UVA_A")

pop_A_U <- c("AFU_A", "UVA_A")

```

```{R}

# Subset the genind object to keep only the selected populations

##Gulf of Panama

# Create a logical vector indicating which individuals belong to the selected populations
pop_subset_S_M <- genind_reef@pop %in% pop_S_M

# Subset the genind object to keep only the selected populations
genind_S_M <- genind_reef[pop_subset_S_M, ]
```

```{r}

###sanity check

head(genind_S_M)

genind_S_M@pop

```

```{r}
# Desired order of labels
lab_order = c("CON_A","MOG_A","SAB_A","AFU_A","BAD_A","UVA_A")

# Change order of rows and cols
fst.mat = as.matrix(coral_reef_fst)
fst.mat1 = fst.mat[lab_order, ]
fst.mat2 = fst.mat1[, lab_order]

# Create a data.frame
ind = which(upper.tri(fst.mat2), arr.ind = TRUE)
fst.df = data.frame(Site1 = dimnames(fst.mat2)[[2]][ind[,2]],
                    Site2 = dimnames(fst.mat2)[[1]][ind[,1]],
                    Fst = fst.mat2[ ind ])

##Keep the order of the levels in the data.frame for plotting 
fst.df$Site1 = factor(fst.df$Site1, levels = unique(fst.df$Site1))
fst.df$Site2 = factor(fst.df$Site2, levels = unique(fst.df$Site2))

# Convert minus values to zero
fst.df$Fst[fst.df$Fst < 0] = 0

# Print data.frame summary
fst.df %>% str
## 'data.frame':    15 obs. of  3 variables:
 # $ Site1: Factor w/ 5 levels "Brd","Pad","Vig",..: 1 2 2 3 3 3 4 4 4 4 ...
 # $ Site2: Factor w/ 5 levels "Ber","Brd","Pad",..: 1 1 2 1 2 3 1 2 3 4 ...
 # $ Fst  : num  0.007 0.025 0.008 0.064 0.038 0.018 0.174 0.171 0.161 0.112 ...

# Fst italic label
fst.label = expression(italic("F")[ST])

# Extract middle Fst value for gradient argument
mid = max(fst.df$Fst) / 2

# Plot heatmap
ggplot(data = fst.df, aes(x = Site1, y = Site2, fill = Fst))+
  geom_tile(colour = "black")+
  geom_text(aes(label = Fst), color="black", size = 3)+
  scale_fill_gradient2(low = "#1E34D9", mid = "#F57CCF", high = "#FF0070", midpoint = mid, name = fst.label, limits = c(0, max(fst.df$Fst)), breaks = c(0, 0.05, 0.10, 0.15))+
  scale_x_discrete(expand = c(0,0))+
  scale_y_discrete(expand = c(0,0), position = "right")+
  theme(axis.text = element_text(colour = "black", size = 10, face = "bold"),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.position = "right",
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 10)
        )
```


```{r}
##OUTFLANK

library(OutFLANK)

geno <- extract.gt(vcf_prunedLD_uniquenames)
dim(geno)

G <- geno  #we are doing this because we will be running a lot of different things with G, and if we mess up we want to be able to go back to geno

G[geno %in% c("0/0")] <- 0
G[geno  %in% c("0/1")] <- 1
G[geno %in% c("1/1")] <- 2
G[is.na(G)] <- 9
tG <- t(G)
dim(tG)

###Turn pop info into factor

pop.data$loc <- as.factor(pop.data$loc)

head(pop.data)
```
```{r}
fst <- MakeDiploidFSTMat(tG,locusNames=1:ncol(tG),popNames=pop.data$loc)
#> Calculating FSTs, may take a few minutes...
head(fst)
```


```{r}

##Data check, he vs Fst

plot(fst$He, fst$FST)
```

```{r}
###fst vs Fstnocorr

plot(fst$FST, fst$FSTNoCorr)
abline(0,1)
```

```{r}
hist(fst$FST,breaks=50)

summary(fst$FST)
```
```{r}
OF <- OutFLANK(fst,LeftTrimFraction=0.01,RightTrimFraction=0.01,
         Hmin=0.05,NumberOfSamples=2,qthreshold=0.05)

OutFLANKResultsPlotter(OF,withOutliers=T,
                       NoCorr=T,Hmin=0.1,binwidth=0.005,
                       Zoom=F,RightZoomFraction=0.05,titletext=NULL)
```
```{r}
P1 <- pOutlierFinderChiSqNoCorr(fst,Fstbar=OF$FSTNoCorrbar,
                                dfInferred=OF$dfInferred,qthreshold=0.05,Hmin=0.1)

P1

outliers <- P1$OutlierFlag==TRUE #which of the SNPs are outliers?
table(outliers)
```
```{r}
plot(P1$LocusName,P1$FST,xlab="Position",ylab="FST",col=rgb(0,0,0,alpha=0.1))
points(P1$LocusName[outliers],P1$FST[outliers],col="red")
```




###Lightbulb moment, yes of course the behaviour is different, it is not the pop names but NUMBER of pop, see base script:


WC_FST_Diploids_2Alleles<-function(Sample_Mat){
  ##Calculate both Fst and Fst NoCorr at the same time, from WC84
  #Sample Mat has three columns (homo_p,m heterozygotes, and homo_q) and a row for each population
  
  sample_sizes = rowSums(Sample_Mat)
  n_ave = mean(sample_sizes)
  n_pops = nrow(Sample_Mat) #r
  r = n_pops
  n_c = (n_pops*n_ave - sum(sample_sizes^2)/(n_pops*n_ave))/(n_pops-1)
  p_freqs = (Sample_Mat[,1] + Sample_Mat[,2]/2) /sample_sizes
  p_ave = sum(sample_sizes*p_freqs)/(n_ave*n_pops)
  
  s2 = sum(sample_sizes*(p_freqs - p_ave)^2)/((n_pops-1)*n_ave)
  if(s2==0){return(0); break}  
  
  h_freqs = Sample_Mat[,2]/sample_sizes
  h_ave = sum(sample_sizes*h_freqs)/(n_ave*n_pops) 

### OutFLANK estimates the distribution of Fst based on a trimmed sample of Fst's. It assumes that the majority of loci in the center of the distribution are neutral and infers the shape of the distribution of neutral Fst using a trimmed set of loci. Loci with the highest and lowest Fst's are trimmed from the data set before this inference, and the distribution of Fst df/(mean Fst) is assumed to'follow a chi-square distribution. Based on this inferred distribution, each locus is given a q-value based on its quantile in the inferred null'distribution. 

```{r}

###Get outliers from OUTflank's gen position from vcf file 

vcfann <- as.data.frame(getFIX(vcf_prunedLD_uniquenames))

OUTFLANK.outliers <- vcfann[outliers,]

head(OUTFLANK.outliers)

```

```{r}

###Remove NAS in OUTFLANK

OUTFLANK.outliers2 <- OUTFLANK.outliers[, -c(6,7)]

OUTFLANK.outliers2 
```


```{r}
rows_na <- 12:95

OUTFLANK.outliers3 <- OUTFLANK.outliers2[-rows_na, ] 

OUTFLANK.outliers3 
```


```{r}
##Re-run outflank per reef
```


```{r}
##OUTFLANK

library(OutFLANK)

geno <- extract.gt(vcf_prunedLD_uniquenames)
dim(geno)

G <- geno  #we are doing this because we will be running a lot of different things with G, and if we mess up we want to be able to go back to geno

G[geno %in% c("0/0")] <- 0
G[geno  %in% c("0/1")] <- 1
G[geno %in% c("1/1")] <- 2
G[is.na(G)] <- 9
tG <- t(G)
dim(tG)

###Turn pop info into factor

pop.data$reef <- as.factor(pop.data$reef)

head(pop.data)
```
```{r}
fst <- MakeDiploidFSTMat(tG,locusNames=1:ncol(tG),popNames=pop.data$reef)
#> Calculating FSTs, may take a few minutes...
head(fst)
```


```{r}

##Data check, he vs Fst

plot(fst$He, fst$FST)
```

```{r}
###fst vs Fstnocorr

plot(fst$FST, fst$FSTNoCorr)
abline(0,1)
```

```{r}
hist(fst$FST,breaks=50)

summary(fst$FST)
```

```{r}
OF <- OutFLANK(fst,LeftTrimFraction=0.01,RightTrimFraction=0.01,
         Hmin=0.05,NumberOfSamples=2,qthreshold=0.05)

OutFLANKResultsPlotter(OF,withOutliers=T,
                       NoCorr=T,Hmin=0.1,binwidth=0.005,
                       Zoom=F,RightZoomFraction=0.05,titletext=NULL)
```

```{r}
P1 <- pOutlierFinderChiSqNoCorr(fst,Fstbar=OF$FSTNoCorrbar,
                                dfInferred=OF$dfInferred,qthreshold=0.05,Hmin=0.1)

P1

outliers <- P1$OutlierFlag==TRUE #which of the SNPs are outliers?
table(outliers)
```

```{r}

```

